File: src/mutation.rs
use crate::database::*;
use crate::model::*;
use crate::util::*;
use crate::validator::*;
use async_graphql::{Context, Error, Object, Result};
use async_std::sync::{Arc, Mutex};
use chrono::NaiveDate;
use log::debug;
use sqlx::postgres::PgPool;
use sqlx::Acquire;
use tide::sessions::Session;
use tindercrypt::cryptors::RingCryptor;
use uuid::Uuid;

pub struct Mutation;

#[Object]
impl Mutation {
    async fn signin(
        &self,
        ctx: &Context<'_>,
        #[graphql(validator(email))] username: String,
        password: String,
    ) -> Result<UserObject> {
        if let Some(user) = find_user_by_username(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &username,
        )
        .await
        {
            if hash_password(&password) != user.password_hash {
                return Err(Error::new("invalid password"));
            }
            let mut session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
            session
                .insert("key", derive_pbkdf2_key(&username, &password))
                .unwrap();
            session.insert("id", user.id).unwrap();
            Ok(UserObject::from_db(user))
        } else {
            Err(Error::new("unknown username"))
        }
    }

    async fn signout(&self, ctx: &Context<'_>) -> Result<bool> {
        ctx.data_unchecked::<Arc<Mutex<Session>>>()
            .lock()
            .await
            .destroy();
        Ok(true)
    }

    async fn signup(
        &self,
        ctx: &Context<'_>,
        #[graphql(validator(custom = "EmailValidator::new()"))] username: String,
        #[graphql(validator(custom = "PasswordValidator::new()"))] password: String,
    ) -> Result<UserObject> {
        let mut connection = ctx.data_unchecked::<PgPool>().acquire().await.unwrap();
        if find_user_by_username(&mut connection, &username)
            .await
            .is_none()
        {
            create_user(&mut connection, &username, &hash_password(&password)).await;
            self.signin(ctx, username, password).await
        } else {
            Err(Error::new("username already exists"))
        }
    }

    async fn update_username(
        &self,
        ctx: &Context<'_>,
        #[graphql(validator(email))] new_username: String,
        password: String,
    ) -> Result<bool> {
        debug!("updating username");
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            let mut connection = ctx.data_unchecked::<PgPool>().acquire().await.unwrap();
            let cryptor = ctx.data_unchecked::<RingCryptor>();
            let key = session.get::<String>("key").unwrap();
            let user = find_user_by_id(&mut connection, &id).await;
            if hash_password(&password) != user.password_hash {
                return Err(Error::new("invalid password"));
            }
            if find_user_by_username(
                &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
                &new_username,
            )
            .await
            .is_none()
            {
                let mut transaction = Acquire::begin(&mut connection)
                    .await
                    .expect("Failed to begin transaction");
                reencrypt_all_data(
                    &mut transaction,
                    cryptor,
                    &key,
                    &id,
                    &new_username,
                    &password,
                )
                .await?;
                update_user_username(&mut transaction, &id, &new_username).await;
                transaction
                    .commit()
                    .await
                    .expect("Failed to commit transaction");
                Ok(true)
            } else {
                Err(Error::new("username already exists"))
            }
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn update_password(
        &self,
        ctx: &Context<'_>,
        password: String,
        #[graphql(validator(custom = "PasswordValidator::new()"))] new_password: String,
    ) -> Result<bool> {
        debug!("updating password");
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            let mut connection = ctx.data_unchecked::<PgPool>().acquire().await.unwrap();
            let cryptor = ctx.data_unchecked::<RingCryptor>();
            let key = session.get::<String>("key").unwrap();
            let user = find_user_by_id(&mut connection, &id).await;
            if hash_password(&password) != user.password_hash {
                return Err(Error::new("invalid password"));
            }
            let mut transaction = Acquire::begin(&mut connection)
                .await
                .expect("Failed to begin transaction");
            reencrypt_all_data(
                &mut transaction,
                cryptor,
                &key,
                &id,
                &user.username,
                &new_password,
            )
            .await?;
            update_user_password_hash(&mut transaction, &id, &hash_password(&new_password)).await;
            transaction
                .commit()
                .await
                .expect("Failed to commit transaction");
            Ok(true)
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn delete_user(&self, ctx: &Context<'_>) -> Result<bool> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            delete_user_by_id(
                &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
                &id,
            )
            .await;
            self.signout(ctx).await
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn normalize_address(&self, ctx: &Context<'_>, label: String) -> Result<Vec<String>> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        let client = ctx.data_unchecked::<surf::Client>();
        let request = surf::get("https://api-adresse.data.gouv.fr/search/")
            .query(&AddressRequest { q: label })
            .unwrap()
            .build();
        let mut response = client.send(request).await.unwrap();
        let AddressResponse { features } = response.body_json().await.unwrap();
        Ok(features.into_iter().map(|f| f.properties.label).collect())
    }

    async fn create_address(
        &self,
        ctx: &Context<'_>,
        title: String,
        label: String,
        address_type: i16,
    ) -> Result<Uuid> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            let cryptor = ctx.data_unchecked::<RingCryptor>();
            let key = session.get::<String>("key").unwrap();
            Ok(create_address(
                &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
                &encrypt_data(cryptor, &key, &title)?,
                &encrypt_data(cryptor, &key, &label)?,
                address_type,
                &id,
            )
            .await)
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn update_address(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        title: String,
        label: String,
        address_type: i16,
    ) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        let cryptor = ctx.data_unchecked::<RingCryptor>();
        let key = session.get::<String>("key").unwrap();
        Ok(update_address(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
            &encrypt_data(cryptor, &key, &title)?,
            &encrypt_data(cryptor, &key, &label)?,
            address_type,
        )
        .await)
    }

    async fn delete_address(&self, ctx: &Context<'_>, id: Uuid) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(delete_address_by_id(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
        )
        .await)
    }

    async fn create_vehicle(
        &self,
        ctx: &Context<'_>,
        model: String,
        horsepower: i16,
        electric: bool,
        vehicle_type: i16,
    ) -> Result<Uuid> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            Ok(create_vehicle(
                &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
                &encrypt_data(
                    ctx.data_unchecked::<RingCryptor>(),
                    &session.get::<String>("key").unwrap(),
                    &model,
                )?,
                horsepower,
                electric,
                vehicle_type,
                &id,
            )
            .await)
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn update_vehicle(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        model: String,
        horsepower: i16,
        electric: bool,
        vehicle_type: i16,
    ) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(update_vehicle(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
            &encrypt_data(
                ctx.data_unchecked::<RingCryptor>(),
                &session.get::<String>("key").unwrap(),
                &model,
            )?,
            horsepower,
            electric,
            vehicle_type,
        )
        .await)
    }

    async fn delete_vehicle(&self, ctx: &Context<'_>, id: Uuid) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(delete_vehicle_by_id(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
        )
        .await)
    }

    async fn create_driver(
        &self,
        ctx: &Context<'_>,
        name: String,
        limit_distance: bool,
        default_vehicle_id: Option<Uuid>,
        default_from_id: Option<Uuid>,
        default_to_id: Option<Uuid>,
    ) -> Result<Uuid> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if let Some(id) = session.get::<Uuid>("id") {
            Ok(create_driver(
                &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
                &encrypt_data(
                    ctx.data_unchecked::<RingCryptor>(),
                    &session.get::<String>("key").unwrap(),
                    &name,
                )?,
                limit_distance,
                &id,
                default_vehicle_id.as_ref(),
                default_from_id.as_ref(),
                default_to_id.as_ref(),
            )
            .await)
        } else {
            Err(Error::new("not logged in"))
        }
    }

    async fn update_driver(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        name: String,
        limit_distance: bool,
        default_vehicle_id: Option<Uuid>,
        default_from_id: Option<Uuid>,
        default_to_id: Option<Uuid>,
    ) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(update_driver(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
            &encrypt_data(
                ctx.data_unchecked::<RingCryptor>(),
                &session.get::<String>("key").unwrap(),
                &name,
            )?,
            limit_distance,
            default_vehicle_id.as_ref(),
            default_from_id.as_ref(),
            default_to_id.as_ref(),
        )
        .await)
    }

    async fn delete_driver(&self, ctx: &Context<'_>, id: Uuid) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(delete_driver_by_id(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
        )
        .await)
    }

    async fn create_or_update_distance(
        &self,
        ctx: &Context<'_>,
        from_id: Uuid,
        to_id: Uuid,
        meters: i32,
    ) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(create_or_update_distance(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &from_id,
            &to_id,
            meters,
        )
        .await)
    }

    async fn create_journey(
        &self,
        ctx: &Context<'_>,
        from_id: Uuid,
        to_id: Uuid,
        driver_id: Uuid,
        vehicle_id: Uuid,
        date: NaiveDate,
        meters: i32,
        round_trip: bool,
    ) -> Result<Uuid> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(create_journey(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &from_id,
            &to_id,
            &driver_id,
            &vehicle_id,
            &date,
            meters,
            round_trip,
        )
        .await)
    }

    async fn update_journey(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        from_id: Uuid,
        to_id: Uuid,
        driver_id: Uuid,
        vehicle_id: Uuid,
        date: NaiveDate,
        meters: i32,
        round_trip: bool,
    ) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(update_journey(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
            &from_id,
            &to_id,
            &driver_id,
            &vehicle_id,
            &date,
            meters,
            round_trip,
        )
        .await)
    }

    async fn delete_journey(&self, ctx: &Context<'_>, id: Uuid) -> Result<u64> {
        let session = ctx.data_unchecked::<Arc<Mutex<Session>>>().lock().await;
        if session.get::<Uuid>("id").is_none() {
            return Err(Error::new("not logged in"));
        }
        Ok(delete_journey_by_id(
            &mut ctx.data_unchecked::<PgPool>().acquire().await.unwrap(),
            &id,
        )
        .await)
    }
}

File: src/mutation.js
import { gql } from "graphql-request";

import { graphQLClient } from "./graphql.js";
import { user } from "./store.js";

export const passwordLowercaseRegex = /(?=.*[a-z])/;
export const passwordUppercaseRegex = /(?=.*[A-Z])/;
export const passwordNumberRegex = /(?=.*[0-9])/;
export const passwordSpecialCharacterRegex = /(?=.*[ !"#$%&'()*+,./:;<=>?@^_`{|}~[\]\\-])/;
export const passwordLengthRegex = /(?=.{8,24})/;

export async function signin(username, password) {
  const mutation = gql`
    mutation Signin($username: String!, $password: String!) {
      signin(username: $username, password: $password) {
        username
      }
    }
  `;

  const variables = {
    username,
    password,
  };
  const data = await graphQLClient.request(mutation, variables);
  user.set(data.signin);
}

export async function signout() {
  const mutation = gql`
    mutation {
      signout
    }
  `;

  await graphQLClient.request(mutation);
  user.set(undefined);
}

export async function signup(username, password) {
  const mutation = gql`
    mutation Signup($username: String!, $password: String!) {
      signup(username: $username, password: $password) {
        username
      }
    }
  `;

  const variables = {
    username,
    password,
  };
  const data = await graphQLClient.request(mutation, variables);
  user.set(data.signup);
}

export async function updateUsername(newUsername, password) {
  const mutation = gql`
    mutation UpdateUsername($newUsername: String!, $password: String!) {
      updateUsername(newUsername: $newUsername, password: $password)
    }
  `;

  const variables = {
    newUsername,
    password,
  };
  await graphQLClient.request(mutation, variables);
}

export async function updatePassword(password, newPassword) {
  const mutation = gql`
    mutation UpdatePassword($password: String!, $newPassword: String!) {
      updatePassword(password: $password, newPassword: $newPassword)
    }
  `;

  const variables = {
    password,
    newPassword,
  };
  await graphQLClient.request(mutation, variables);
}

export async function createDriver(driver) {
  const mutation = gql`
    mutation CreateDriver(
      $name: String!
      $limitDistance: Boolean!
      $defaultVehicleId: ID
      $defaultFromId: ID
      $defaultToId: ID
    ) {
      createDriver(
        name: $name
        limitDistance: $limitDistance
        defaultVehicleId: $defaultVehicleId
        defaultFromId: $defaultFromId
        defaultToId: $defaultToId
      )
    }
  `;

  await graphQLClient.request(mutation, driver);
}

export async function updateDriver(driver) {
  const mutation = gql`
    mutation UpdateDriver(
      $id: ID!
      $name: String!
      $limitDistance: Boolean!
      $defaultVehicleId: ID
      $defaultFromId: ID
      $defaultToId: ID
    ) {
      updateDriver(
        id: $id
        name: $name
        limitDistance: $limitDistance
        defaultVehicleId: $defaultVehicleId
        defaultFromId: $defaultFromId
        defaultToId: $defaultToId
      )
    }
  `;

  await graphQLClient.request(mutation, driver);
}

export async function deleteDriver(driver) {
  const mutation = gql`
    mutation DeleteDriver($id: ID!) {
      deleteDriver(id: $id)
    }
  `;

  await graphQLClient.request(mutation, driver);
}

export async function createVehicle(vehicle) {
  const mutation = gql`
    mutation CreateVehicle($model: String!, $horsepower: Int!, $electric: Boolean!, $vehicleType: Int!) {
      createVehicle(model: $model, horsepower: $horsepower, electric: $electric, vehicleType: $vehicleType)
    }
  `;

  await graphQLClient.request(mutation, vehicle);
}

export async function updateVehicle(vehicle) {
  const mutation = gql`
    mutation UpdateVehicle($id: ID!, $model: String!, $horsepower: Int!, $electric: Boolean!, $vehicleType: Int!) {
      updateVehicle(id: $id, model: $model, horsepower: $horsepower, electric: $electric, vehicleType: $vehicleType)
    }
  `;

  await graphQLClient.request(mutation, vehicle);
}

export async function deleteVehicle(vehicle) {
  const mutation = gql`
    mutation DeleteVehicle($id: ID!) {
      deleteVehicle(id: $id)
    }
  `;

  await graphQLClient.request(mutation, vehicle);
}

export async function normalizeAddress(label) {
  const mutation = gql`
    mutation NormalizeAddress($label: String!) {
      normalizeAddress(label: $label)
    }
  `;

  const variables = {
    label,
  };
  const data = await graphQLClient.request(mutation, variables);
  return data.normalizeAddress;
}

export async function createAddress(address) {
  const mutation = gql`
    mutation CreateAddress($title: String!, $label: String!, $addressType: Int!) {
      createAddress(title: $title, label: $label, addressType: $addressType)
    }
  `;

  await graphQLClient.request(mutation, address);
}

export async function updateAddress(address) {
  const mutation = gql`
    mutation UpdateAddress($id: ID!, $title: String!, $label: String!, $addressType: Int!) {
      updateAddress(id: $id, title: $title, label: $label, addressType: $addressType)
    }
  `;

  await graphQLClient.request(mutation, address);
}

export async function deleteAddress(address) {
  const mutation = gql`
    mutation DeleteAddress($id: ID!) {
      deleteAddress(id: $id)
    }
  `;

  await graphQLClient.request(mutation, address);
}

export async function createDistance(distance) {
  const mutation = gql`
    mutation CreateDistance($fromId: ID!, $toId: ID!, $meters: Int!) {
      createDistance(fromId: $fromId, toId: $toId, meters: $meters)
    }
  `;

  await graphQLClient.request(mutation, distance);
}

export async function updateDistance(distance) {
  const mutation = gql`
    mutation UpdateDistance($fromId: ID!, $toId: ID!, $meters: Int!) {
      updateDistance(fromId: $fromId, toId: $toId, meters: $meters)
    }
  `;

  await graphQLClient.request(mutation, distance);
}

export async function createJourney(journey) {
  const mutation = gql`
    mutation CreateJourney(
      $driverId: ID!
      $vehicleId: ID!
      $fromId: ID!
      $toId: ID!
      $date: String!
      $meters: Int!
      $roundTrip: Boolean!
    ) {
      createJourney(
        driverId: $driverId
        vehicleId: $vehicleId
        fromId: $fromId
        toId: $toId
        date: $date
        meters: $meters
        roundTrip: $roundTrip
      )
    }
  `;

  await graphQLClient.request(mutation, journey);
}

export async function updateJourney(journey) {
  const mutation = gql`
    mutation UpdateJourney(
      $id: ID!
      $driverId: ID!
      $vehicleId: ID!
      $fromId: ID!
      $toId: ID!
      $date: String!
      $meters: Int!
      $roundTrip: Boolean!
    ) {
      updateJourney(
        id: $id
        driverId: $driverId
        vehicleId: $vehicleId
        fromId: $fromId
        toId: $toId
        date: $date
        meters: $meters
        roundTrip: $roundTrip
      )
    }
  `;

  await graphQLClient.request(mutation, journey);
}

export async function deleteJourney(journey) {
  const mutation = gql`
    mutation DeleteJourney($id: ID!) {
      deleteJourney(id: $id)
    }
  `;

  await graphQLClient.request(mutation, journey);
}
